### PR TITLE
use getDeclaredConstructor().newInstance() in ConfigurationSetup

### PR DESCRIPTION
**Pull request by vorburger**
https://errorprone.info/bugpattern/ClassNewInstance suggests this

and introduce new InvocationTargetError

### CODE REVIEW

**Comment by gsmet** [comment]
FYI, the CI error was a timeout of our JBoss Nexus. Now the funny thing is that you need to force push somehow to get CI to run again and push the status properly to GitHub.

**Comment by vorburger** [comment]
> you need to force push somehow to get CI to run again and push the status properly to GitHub.

done (`git rebase master`)

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: So I would have merged this one already... if the "let's add lambdas everywhere" part was out of it.

We are very cautious about where we add lambdas and it requires thinking twice about it.

Could you make your PR only about the proposed change?

**Review by gsmet — CHANGES_REQUESTED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (dmlloyd) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: This isn't the right handling for this.  The cause of the ITE should be extracted and rethrown, wrapping with `UndeclaredThrowableException` in the event that it's a `Throwable` that isn't a `RuntimeException` or `Error`.

File: core/deployment/src/main/java/io/quarkus/deployment/util/InvocationTargetError.java
Comment (dmlloyd) [inline core/deployment/src/main/java/io/quarkus/deployment/util/InvocationTargetError.java]: We don't need this; see above comment.

**Review by dmlloyd — CHANGES_REQUESTED** [review]

**Comment by gsmet** [comment]
@vorburger do you plan to update this PR or should we close it?

**Comment by vorburger** [comment]
> @vorburger do you plan to update this PR or should we close it?

@gsmet thanks for the ping, I'll try to do this today!

**Comment by dmlloyd** [comment]
Could it wait until after #1478 is merged?

**Comment by vorburger** [comment]
> Could it wait until after #1478 is merged?

sure thing, I'll hold off... saw that `ConfigurationSetup` significantly changes in #1478, so make sense.

**Comment by dmlloyd** [comment]
Thanks!

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (vorburger) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Sorry, I pre-configure some Auto Save Actions, which caused this. Have just disabled, for future.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (vorburger) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: OK, see if the new version (force pushed) is better? There is a hiccup though... see the TODO: It's not possible (AFAIK) to re-throw the `InvocationTargetException` as either a `RuntimeException` OR an `Error`, with those `toError`method signatures _returning_ instead of _throwing_. Do you want a PR with that changed? Or just forget about using `ReflectUtil` and inline it for that case,with x2 `throw`?

File: core/deployment/src/main/java/io/quarkus/deployment/util/InvocationTargetError.java
Comment (vorburger) [reply core/deployment/src/main/java/io/quarkus/deployment/util/InvocationTargetError.java]: sure, it's gone.

**Comment by vorburger** [comment]
@dmlloyd re-review this?

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (dmlloyd) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Right it should be throwing the *cause*, like this:

```java
// ...
} catch (InvocationTargetException e) {
    try {
        throw e.getCause();
    } catch (RuntimeException | Error e2) {
        throw e2;
    } catch (Throwable t) {
        throw new UndeclaredThrowableException(t);
    }
}
```

It's wordy but it is the "most" correct.

Comment (vorburger) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: done

**Review by dmlloyd — APPROVED** [review]

**Comment by dmlloyd** [comment]
Looks OK for me now, @gsmet are you OK with this now?

**Comment by vorburger** [comment]
`RestClientITCase` failed in CI, and is reproducible locally with `../../mvnw -Dnative=true clean install` (NB: IT) in `integration-tests/main` (after `../../mvnw clean install` in `core/deployment` on branch with this change)...

```
[ERROR] Failures: 
[ERROR]   RestClientITCase>RestClientTestCase.testMicroprofileCdiClientHeaderPassing:88 1 expectation failed.
JSON path header-name doesn't match.
Expected: some-not-at-all-random-header-value
  Actual: null

[ERROR]   RestClientITCase>RestClientTestCase.testMicroprofileClientHeaderPassing:98 1 expectation failed.
JSON path header-name doesn't match.
Expected: some-not-at-all-random-header-value
  Actual: null
```

**Comment by vorburger** [comment]
> `RestClientITCase` failed in CI

This is not related to this change, but a more general problem on current `master` - see #1565.

**Comment by vorburger** [comment]
> Looks OK for me now, @gsmet are you OK with this now?

@gsmet merge?

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Merged, thanks!

**Pull request merged**
